tests/MongoDB.Driver.Tests/CustomServerSelectorTests.cs

```diff
@@ -0,0 +1,86 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Servers;
+using Xunit;
+
+namespace MongoDB.Driver.Tests
+{
+    public class CustomServerSelectorTests
+    {
+        [Fact]
+        public void Should_call_custom_server_selector()
+        {
+            var eventCapturer = new EventCapturer()
+                .Capture<ClusterSelectingServerEvent>()
+                .Capture<ClusterSelectedServerEvent>();
+            var customServerSelector = new CustomServerSelector();
+            var client = DriverTestConfiguration.CreateDisposableClient(
+                clientSettings =>
+                    clientSettings.ClusterConfigurator =
+                        c =>
+                        {
+                            c.ConfigureCluster(
+                                s =>
+                                    new ClusterSettings(
+                                        postServerSelector: customServerSelector));
+                            c.Subscribe(eventCapturer);
+                        });
+
+            var collection = client
+                .GetDatabase(DriverTestConfiguration.DatabaseNamespace.DatabaseName)
+                .GetCollection<BsonDocument>(DriverTestConfiguration.CollectionNamespace.CollectionName)
+                .WithReadPreference(ReadPreference.Nearest);
+
+            for (int i = 0; i < 10; i++)
+            {
+                customServerSelector.CustomSelectorWasCalled = false;
+                eventCapturer.Clear();
+
+                collection.CountDocuments(new BsonDocument());
+
+                customServerSelector.CustomSelectorWasCalled.Should().BeTrue();
+                eventCapturer.Next().Should().BeOfType<ClusterSelectingServerEvent>();
+                eventCapturer.Next().Should().BeOfType<ClusterSelectedServerEvent>();
+            }
+        }
+
+        // nested types
+        private class CustomServerSelector : IServerSelector
+        {
+            public bool CustomSelectorWasCalled { get; set; }
+
+            public IEnumerable<ServerDescription> SelectServers(ClusterDescription cluster, IEnumerable<ServerDescription> servers)
+            {
+                CustomSelectorWasCalled = true;
+                var server = servers.FirstOrDefault(x => ((DnsEndPoint)x.EndPoint).Port == 27017);
+
+                return server != null
+                    ? new ServerDescription[1] {server}
+                    : new ServerDescription[0];
```

Comment:
I think lines 78-82 can be simplified to:

    return servers.Where(x => ((DnsEndPoint)x.EndPoint).Port == 27017).Take(1);

Reply:
Or even further simplified to:

    return servers;

Not sure why filtering is relevant to the tests. All we need to know is whether the custom server selector was called or not.